### ISSUE TITLE
Support for JDK 11

### ISSUE THREAD

**Issue by ahamana** [payara/ecosystem-eclipse-plugin#11]
Does the current version 1.0 support for JDK 11?
When the JDK 11 path is specified on the Payara Server settings dialog, the message "Java Development Kit (JDK) is required rather than a JRE" is shown. 

![payara](https://user-images.githubusercontent.com/21101083/58756368-910e4000-8532-11e9-8411-2015b8d8a690.png)

**Comment by jGauravGupta** [comment]
Hi @ahamana ,

Payara Eclipse plugin v1.1 will be released soon with JDK 11 support.

**Comment by ahamana** [comment]
@jGauravGupta 
Many thanks for your reply.
I look forward to the new release.

**Comment by carlbradwell17** [comment]
+1 on this, i was about to raise this issue myself.

**Comment by carlbradwell17** [comment]
Hi,
I've build and tried the 1.0.0.201906280050 version of this plugin.
I can now add the JDK11 server runtime environment in eclipse.

However, I cannot get the server to start correctly when starting it in Eclipse.

In the VM arguements, I have added:
`--add-modules java.se --add-exports java.base/jdk.internal.ref=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.nio=ALL-UNNAMED --add-opens java.base/sun.nio.ch=ALL-UNNAMED --add-opens java.management/sun.management=ALL-UNNAMED --add-opens jdk.management/com.sun.management.internal=ALL-UNNAMED --add-opens java.base/jdk.internal.loader=ALL-UNNAMED`

The server starts, but appears as stuck on 53% in eclipse. It timesout and shows in eclipse as not started, but it is.
The startup log shows this:
```
[2019-07-02T09:38:33.396+0100] [Payara 5.192] [INFO] [NCLS-CORE-00017] [javax.enterprise.system.core] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1562056713396] [levelValue: 800] [[
  Payara Server  5.192 #badassfish (115) startup time : Felix (44,267ms), startup services(6,093ms), total(50,360ms)]]

[2019-07-02T09:38:33.400+0100] [Payara 5.192] [INFO] [] [fish.payara.nucleus.notification.NotificationService] [tid: _ThreadID=1 _ThreadName=main] [timeMillis: 1562056713400] [levelValue: 800] [[
  Payara Notification Service bootstrapped with configuration: NotificationExecutionOptions{enabled=true, notifierConfigurationExecutionOptionsList={LOG=LogNotifierConfigurationExecutionOptions{useSeparateLogFile=false} NotifierConfigurationExecutionOptions{notifierType=LOG, enabled=true, noisy=true}}}]]

[2019-07-02T09:38:33.640+0100] [Payara 5.192] [INFO] [NCLS-CORE-00101] [javax.enterprise.system.core] [tid: _ThreadID=28 _ThreadName=RunLevelControllerThread-1562056707307] [timeMillis: 1562056713640] [levelValue: 800] [[
  Network Listener JMS_PROXY_default_JMS_host started in: 2ms - bound to [/0.0.0.0:7676]]]

[2019-07-02T09:38:33.756+0100] [Payara 5.192] [INFO] [NCLS-JMX-00005] [javax.enterprise.system.jmx] [tid: _ThreadID=132 _ThreadName=Thread-9] [timeMillis: 1562056713756] [levelValue: 800] [[
  JMXStartupService has started JMXConnector on JMXService URL service:jmx:rmi://0.0.0.0:8686/jndi/rmi://0.0.0.0:8686/jmxrmi]]

[2019-07-02T09:38:33.842+0100] [Payara 5.192] [INFO] [] [org.jvnet.hk2.osgiadapter] [tid: _ThreadID=23 _ThreadName=FelixStartLevel] [timeMillis: 1562056713842] [levelValue: 800] [[
  Skipping registration of inhabitant for service reference [org.osgi.service.metatype.MetaTypeProvider] as the service object could not be obtained.]]
```

So at this stage, does not seem to be able to fully work with JDK 11.

**Comment by CarlosMOGoncalves** [comment]
Let me add my two cents to that. 

When you use, for example these:
`--add-modules java.se`
`--add-exports java.base/jdk.internal.ref=ALL-UNNAMED`
`--add-opens java.base/java.nio=ALL-UNNAMED`

it will start on Eclipse just fine. Those seem like the right arguments for the JVM args.

However, if you run that with `asadmin start-domain` it will refuse to start stating that the server exited with an error code of 1 `Unrecognized option: --add-opens java.base/java.nio=ALL-UNNAMED` .

Now if you add an `=` between those JVM args it will start just fine with `asadmin start-domain`. 

Like this:
`--add-modules=java.se`
`--add-exports=java.base/jdk.internal.ref=ALL-UNNAMED`
`--add-opens=java.base/java.nio=ALL-UNNAMED`

This does not seem like standard JVM args notation, however it is exactly like this that the standard domain1 comes configured in Payara 5.192.

However, if you try to run it in Eclipse with Payara Tools it will not start and shall be stuck. This seems to point that lauched from Eclipse it will only accept JVM args **without** = but from `asadmin start-domain` will only accept JVM args with =.

I am speculating that `start-domain` script might treat those args differently but it does not seem to do anything special other than launching a jar.

**Note**: Eclipse with Payara Tools and NetBeans 11.1 with Payara server integration work perfectly without the = in the JVM args. Intellij doesn't... because it launches start-domain. How do I know this? Well, my team uses all those IDEs and I have just performed a GF -> Payara migration, JDK8 -> OJDK11.

**Comment by carlbradwell17** [comment]
Hi,
I've tried it with just the JVM options mentioned by @CarlosMOGoncalves , and got no further. It starts fine when using asadmin, but never in eclipse. Switching back to java 1.8 and it starts straight away.
Always gets stuck at 53%
`[2019-07-26T07:19:07.167+0000] [Payara 5.192] [INFO] [] [org.jvnet.hk2.osgiadapter] [tid: _ThreadID=23 _ThreadName=FelixStartLevel] [timeMillis: 1564125547167] [levelValue: 800] [[
  Skipping registration of inhabitant for service reference [org.osgi.service.metatype.MetaTypeProvider] as the service object could not be obtained.]]`

**Comment by CarlosMOGoncalves** [comment]
One other thing, have you created a password for your domain? I found this problem when I first added a password to the domain and then Payara Tools did get stuck at that percentage (which leads me to believe that if you apply a password it will defer back to `start-domain`).

If you did apply a password, could you kindly try a fresh install (download it again, apply your domain, do not set a domain password, use JVM options WITHOUT the = and try to start it with Eclipse Payara Tools) and keep this thread posted on whether it worked or not? This is more of a diagnostic thing more than everything else.

**Comment by carlbradwell17** [comment]
I have used a fresh download of Payara 5.192 and made no changes to the domain., still same result.
I have tried it with just the 3 VM arguments you suggested, although the others I mentioned in my first comment seem to be required for hazelcast to start without warnings.

I have also tried deleting the startserv.bat file to make sure it is not using that.

The server is actually started, as I can see it in the console logs and can access the admin portal in a browser. The default timeout in eclipse is still set at 240 seconds. Even when eclipse reports "Unable to start server on time", it is still accessible. I have increased the timeout and even leaving it 10 minutes it never gets past the point mentioned previously.

**Comment by CarlosMOGoncalves** [comment]
No no, friend. Those 3 were just an example. You need all of them with the same notation, something like this:

`<jvm-options>[9|]--add-opens java.base/jdk.internal.loader=ALL-UNNAMED</jvm-options>`
`<jvm-options>[9|]--add-opens jdk.management/com.sun.management.internal=ALL-UNNAMED</jvm-options>`
`<jvm-options>[9|]--add-exports java.base/jdk.internal.ref=ALL-UNNAMED</jvm-options>`
`<jvm-options>[9|]--add-opens java.base/java.lang=ALL-UNNAMED</jvm-options>`
`<jvm-options>[9|]--add-opens java.base/java.nio=ALL-UNNAMED</jvm-options>`
`<jvm-options>[9|]--add-opens java.base/sun.nio.ch=ALL-UNNAMED</jvm-options>`
`<jvm-options>[9|]--add-opens java.management/sun.management=ALL-UNNAMED</jvm-options>`
`<jvm-options>[9|]--add-opens java.base/sun.net.www.protocol.jrt=ALL-UNNAMED</jvm-options>`

With these, I am able to start it perfectly on Eclipse Payara Tools, but not with `start-domain`. 

What you mentioned about Hazelcast happened simply because, again, starting on Payara Tools with the default JVM options from the domain counts as if it cannot find them properly, so it complains that the Java 9+  modular structure prevents Hazelcast from finding some classes it needs.

Anyway, I think the developers of Payara Tools now have enough information about this to work with.

**Comment by jGauravGupta** [comment]
`jvm-options` are already parsed and added by Payara Tools [@ServerTasks.java#L151](https://github.com/payara/ecosystem-eclipse-plugin/blob/1.0.0.201906280050/plugins/org.eclipse.payara.tools/src/org/eclipse/payara/tools/sdk/server/ServerTasks.java#L151) from `domain.xml` before starting the server so adding VM arguments explicitly should not be required.

//cc @arjantijms

**Comment by CarlosMOGoncalves** [comment]
That is correct, I was mentioning that these jvm options should be used in `domain.xml`.

What I found out, related to this thread, is that if you keep the default ones that come straight from Payara 5.192 (with an equals sign between the option, just like I typed in my first post), the server will start just fine on the command-line using `asadmin start-domain`. However, from Payara Tools it will complain a lot about the missing JVM options. 

If however you remove the equals sign it will start perfectly on the Payara Tools, but not from the command-line. That might have something to do with it.

**Comment by arjantijms** [comment]
I'll take a more in-depth look at this soon then. For now, I'm fairly sure we tested this plug-in version on both Linux (Ubuntu 19.04), macOS (Mojave 10.14.5) and Windows, with out of the box Payara versions (5.192).

I *should* work. See also the screenshots I took here:

https://blog.payara.fish/payara-tools-1.0-2019-06-new-and-noteworthy

**Comment by carlbradwell17** [comment]
OK there was some confusion over the VM arguments, as @CarlosMOGoncalves suggested, they should be modified in domain.xml not added as VM arguments in Eclipse as I was previously doing ( I picked that up from another post somewhere).
I have managed to get Payara to start in Eclipse now doing the following:

- Start with a fresh copy of the default domain1 and make no changes to it
- Edit domain1\config\domain.xml
- find and replace "add-opens=" with "add-opens " (there are 14 occurrences)
- find and replace "add-exports=" with "add-exports " (there are 2 occurrences)

When I did this in my custom domain it still did not work. As @CarlosMOGoncalves suggested it could be the password, but unsetting it in that domain still did not work.
So I am going to go through that domain template and add each change versus the default domain back in one at a time.
